README.md
# ashlar


## Developing

**Ashlar is under heavy development in tandem with
[DRIVER](https://github.com/WorldBank-Transport/DRIVER); currently the best way to develop
on Ashlar is to set up a DRIVER VM and use that.**


## Testing

TODO: Add testing how-to


CHANGELOG.md
# Changelog

## 2.0dev

Changes we're making en route to version 2.0. When we cut a release, this
section will form the changelog for v2.0.

- Removed `weather` and `light` fields from the `Record` data model
  ([#85](https://github.com/azavea/ashlar/pull/85))
- Updated Django to version 1.11 and dropped support for version 1.8
  ([#84](https://github.com/azavea/ashlar/pull/84))
- Added support for Python3, meaning that Ashlar supports Python versions 2.7,
  3.4, 3.5, and 3.6 ([#82](https://github.com/azavea/ashlar/pull/82))

## 1.0 

Initial release of Ashlar.


ashlar/serializers.py
import logging
from django.db import transaction
from django.conf import settings

import datetime
import json
import pytz
import requests

from rest_framework import serializers
from rest_framework.serializers import ModelSerializer
from rest_framework_gis.serializers import GeoFeatureModelSerializer, GeoModelSerializer

from ashlar.models import Boundary, BoundaryPolygon, Record, RecordType, RecordSchema
from ashlar.serializer_fields import JsonBField, JsonSchemaField, GeomBBoxField

logger = logging.getLogger(__name__)


class RecordSerializer(GeoModelSerializer):

    data = JsonBField()

    class Meta:
        model = Record
        fields = '__all__'
        read_only_fields = ('uuid',)


class RecordTypeSerializer(ModelSerializer):

    current_schema = serializers.SerializerMethodField()

    class Meta:
        model = RecordType
        fields = '__all__'

    def get_current_schema(self, obj):
        current_schema = obj.get_current_schema()
        uuid = None
        if current_schema:
            uuid = str(current_schema.uuid)
        return uuid


class SchemaSerializer(ModelSerializer):
    """Base class for serializers of subclasses of models.SchemaModel"""
    schema = JsonSchemaField()


class RecordSchemaSerializer(SchemaSerializer):

    def create(self, validated_data):
        """Creates new schema or creates new version and updates next_version of previous"""
        if validated_data['version'] > 1:  # Viewset's get_serializer() will always add 'version'
            with transaction.atomic():
                current = RecordSchema.objects.get(record_type=validated_data['record_type'],
                                                   next_version=None)
                new = RecordSchema.objects.create(**validated_data)
                current.next_version = new
                current.save()
        elif validated_data['version'] == 1:  # New record_type
            new = RecordSchema.objects.create(**validated_data)
        else:
            raise serializers.ValidationError('Schema version could not be determined')
        return new

    class Meta:
        model = RecordSchema
        fields = '__all__'
        read_only_fields = ('uuid', 'next_version')


class BoundaryPolygonSerializer(GeoFeatureModelSerializer):

    data = JsonBField()

    class Meta:
        model = BoundaryPolygon
        geo_field = 'geom'
        id_field = 'uuid'
        exclude = ('boundary',)


class BoundaryPolygonNoGeomSerializer(ModelSerializer):
    """Serialize a BoundaryPolygon without any geometry info"""
    data = JsonBField()
    bbox = GeomBBoxField(source='geom')

    class Meta:
        model = BoundaryPolygon
        exclude = ('geom',)


class BoundarySerializer(GeoModelSerializer):

    label = serializers.CharField(max_length=128, allow_blank=False)
    color = serializers.CharField(max_length=64, required=False)
    display_field = serializers.CharField(max_length=10, allow_blank=True, required=False)
    data_fields = JsonBField(read_only=True, allow_null=True)
    errors = JsonBField(read_only=True, allow_null=True)

    def create(self, validated_data):
        boundary = super(BoundarySerializer, self).create(validated_data)
        boundary.load_shapefile()
        return boundary

    class Meta:
        model = Boundary
        # These meta read_only/exclude settings only apply to the fields the ModelSerializer
        # instantiates for you by default. If you override a field manually, you need to override
        # all settings there.
        # e.g. adding 'errors' to this tuple has no effect, since we manually define the errors
        # field above
        read_only_fields = ('uuid', 'status',)
        fields = '__all__'


app/api/serializer_fields.py
from django.core.exceptions import ValidationError

from rest_framework.fields import Field

import jsonschema
from jsonschema.exceptions import SchemaError

class JsonBField(Field):
    """ Custom serializer class for JsonB

    Do no transformations, as we always want python dicts
    Ensure top level is an object or array

    """
    type_name = 'JsonBField'

    def to_representation(self, value):
        return value

    def to_internal_value(self, value):
        if isinstance(value, dict) or isinstance(value, list):
            return value
        else:
            raise ValidationError('Array or object required')


class JsonSchemaField(JsonBField):
    """ Extend JsonBField to validate on conversion """

    type_name = 'JsonSchemaField'

    def to_internal_value(self, value):
        try:
            jsonschema.Draft4Validator.check_schema(value)
            return value
        except SchemaError as e:
            raise ValidationError('Invalid schema: {}'.format(e.message))


ashlar/migrations/0011_auto_20150505_1941.py
# -*- coding: utf-8 -*-
from __future__ import unicode_literals

from django.db import models, migrations
import uuid


class Migration(migrations.Migration):

    dependencies = [
        ('ashlar', '0010_auto_20150505_1940'),
    ]

    operations = [
        migrations.CreateModel(
            name='RecordType',
            fields=[
                ('uuid', models.UUIDField(default=uuid.uuid4, serialize=False, editable=False, primary_key=True)),
                ('created', models.DateTimeField(auto_now_add=True)),
                ('modified', models.DateTimeField(auto_now=True)),
                ('label', models.CharField(max_length=64)),
                ('plural_label', models.CharField(max_length=64)),
                ('description', models.TextField(null=True, blank=True)),
            ],
            options={
                'abstract': False,
            },
        ),
        # NOTE: makemigrations attempted this operation as an AlterField, but text cannot be
        #       converted to fk via AlterField. So, it was replaced with a
        #       RemoveField + AddField operation
        migrations.RemoveField(
            model_name='recordschema',
            name='record_type',
        ),
        migrations.AddField(
            model_name='recordschema',
            name='record_type',
            field=models.ForeignKey(related_name='schemas', to='ashlar.RecordType'),
        ),
    ]


ashlar/models.py
import os
import shutil
import uuid

from django.conf import settings
from django.contrib.gis.db import models
from django.contrib.gis.gdal import DataSource as GDALDataSource
from djsonb import fields as jsb
from django.core.validators import MinLengthValidator

import jsonschema

from ashlar.imports.shapefile import (extract_zip_to_temp_dir,
                                      get_shapefiles_in_dir,
                                      make_multipolygon)


class AshlarModel(models.Model):
    uuid = models.UUIDField(primary_key=True, default=uuid.uuid4, editable=False)
    created = models.DateTimeField(auto_now_add=True)
    modified = models.DateTimeField(auto_now=True)

    class Meta(object):
        abstract = True


class SchemaModel(AshlarModel):
    version = models.PositiveIntegerField()
    schema = jsb.JsonBField()
    next_version = models.OneToOneField('self', related_name='previous_version', null=True,
                                        editable=False)

    class Meta(object):
        abstract = True

    def validate_json(self, json_dict):
        """Validates a JSON-like dictionary against this object's schema

        :param json_dict: Python dict representing json to be validated against self.schema
        :return: None if validation succeeds; jsonschema.exceptions.ValidationError if failure
                 (or jsonschema.exceptions.SchemaError if the schema is invalid)
        """
        return jsonschema.validate(json_dict, self.schema)

    @classmethod
    def validate_schema(self, schema):
        """Validates that this object's schema is a valid JSON-Schema schema
        :param schema: Python dict representing json schema that should be checked
        :return: None if schema validates; raises jsonschema.exceptions.SchemaError
            if schema is invalid
        """
        jsonschema.Draft4Validator.check_schema(schema)


class Record(AshlarModel):
    """Spatiotemporal records -- e.g. Loch Ness Monster sightings, crime events, etc."""
    occurred_from = models.DateTimeField()
    occurred_to = models.DateTimeField()

    geom = models.PointField(srid=settings.ASHLAR['SRID'])
    location_text = models.CharField(max_length=200, null=True, blank=True)
    city = models.CharField(max_length=50, null=True, blank=True)
    city_district = models.CharField(max_length=50, null=True, blank=True)
    county = models.CharField(max_length=50, null=True, blank=True)
    neighborhood = models.CharField(max_length=50, null=True, blank=True)
    road = models.CharField(max_length=200, null=True, blank=True)
    state = models.CharField(max_length=50, null=True, blank=True)

    schema = models.ForeignKey('RecordSchema')
    data = jsb.JsonBField()

    archived = models.BooleanField(default=False)

    objects = models.GeoManager()

    class Meta(object):
        ordering = ('-created',)


class RecordType(AshlarModel):
    """ Store extra information for a given RecordType, associated schemas in RecordSchema """
    label = models.CharField(max_length=64)
    plural_label = models.CharField(max_length=64)
    description = models.TextField(blank=True, null=True)
    active = models.BooleanField(default=True)

    def get_current_schema(self):
        schemas = self.schemas.order_by('-version')
        return schemas[0] if len(schemas) > 0 else None


class RecordSchema(SchemaModel):
    """Schemas for spatiotemporal records"""
    record_type = models.ForeignKey('RecordType', related_name='schemas')

    class Meta(object):
        unique_together = (('record_type', 'version'),)


class Boundary(AshlarModel):
    """ MultiPolygon objects which contain related geometries for filtering/querying """

    class StatusTypes(object):
        PENDING = 'PENDING'
        PROCESSING = 'PROCESSING'
        ERROR = 'ERROR'
        WARNING = 'WARNING'
        COMPLETE = 'COMPLETE'
        CHOICES = (
            (PENDING, 'Pending'),
            (PROCESSING, 'Processing'),
            (WARNING, 'Warning'),
            (ERROR, 'Error'),
            (COMPLETE, 'Complete'),
        )

    status = models.CharField(max_length=10,
                              choices=StatusTypes.CHOICES,
                              default=StatusTypes.PENDING)
    label = models.CharField(max_length=128, unique=True, validators=[MinLengthValidator(3)])
    # Store any valid css color string
    color = models.CharField(max_length=64, default='blue')
    display_field = models.CharField(max_length=10, blank=True, null=True)
    data_fields = jsb.JsonBField(blank=True, null=True)
    errors = jsb.JsonBField(blank=True, null=True)
    source_file = models.FileField(upload_to='boundaries/%Y/%m/%d')

    def load_shapefile(self):
        """ Validate the shapefile saved on disk and load into db """
        self.status = self.StatusTypes.PROCESSING
        self.save()

        try:
            temp_dir = extract_zip_to_temp_dir(self.source_file)
            shapefiles = get_shapefiles_in_dir(temp_dir)

            if len(shapefiles) != 1:
                raise ValueError('Exactly one shapefile (.shp) required')

            shapefile_path = os.path.join(temp_dir, shapefiles[0])
            shape_datasource = GDALDataSource(shapefile_path)
            if len(shape_datasource) > 1:
                raise ValueError('Shapefile must have exactly one layer')

            boundary_layer = shape_datasource[0]
            if boundary_layer.srs is None:
                raise ValueError('Shapefile must include a .prj file')
            self.data_fields = boundary_layer.fields
            for feature in boundary_layer:
                feature.geom.transform(settings.ASHLAR['SRID'])
                geometry = make_multipolygon(feature.geom)
                data = {field: feature.get(field) for field in self.data_fields}
                self.polygons.create(geom=geometry, data=data)

            self.status = self.StatusTypes.COMPLETE
            self.save()
        except Exception as e:
            if self.errors is None:
                self.errors = {}
            self.errors['message'] = str(e)
            # Relabel geography to allow saving a valid shapefile in this namespace
            self.label = self.label + '_' + str(uuid.uuid4())
            self.status = self.StatusTypes.ERROR
            self.save()
        finally:
            shutil.rmtree(temp_dir, ignore_errors=True)


class BoundaryPolygon(AshlarModel):
    """ Individual boundaries and associated data for each geom in a BoundaryUpload """

    boundary = models.ForeignKey('Boundary', related_name='polygons', null=True)
    data = jsb.JsonBField()
    geom = models.MultiPolygonField(srid=settings.ASHLAR['SRID'])

    objects = models.GeoManager()
